File: sorting .cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;

class QuickSort {
private:
    int comparisons;
    int swaps;

    int partition(vector<int>& arr, int low, int high, bool randomized) {
        if (randomized) {
            int randomIndex = low + rand() % (high - low + 1);
            swap(arr[randomIndex], arr[high]);
            swaps++;
        }

        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            comparisons++;
            if (arr[j] <= pivot) {
                i++;
                swap(arr[i], arr[j]);
                swaps++;
            }
        }

        swap(arr[i + 1], arr[high]);
        swaps++;
        return i + 1;
    }

    void quickSort(vector<int>& arr, int low, int high, bool randomized) {
        if (low < high) {
            int pi = partition(arr, low, high, randomized);
            quickSort(arr, low, pi - 1, randomized);
            quickSort(arr, pi + 1, high, randomized);
        }
    }

public:
    void sortAndShow(vector<int> arr, bool randomized) {
        comparisons = 0;
        swaps = 0;

        clock_t start = clock();
        quickSort(arr, 0, arr.size() - 1, randomized);
        clock_t end = clock();

        double timeTaken = double(end - start) / CLOCKS_PER_SEC;

        cout << "\n" << (randomized ? "RANDOMIZED" : "DETERMINISTIC") << " QUICKSORT\n";
        cout << "Sorted array: [";
        for (size_t i = 0; i < arr.size(); i++) {
            cout << arr[i];
            if (i < arr.size() - 1) cout << ", ";
        }
        cout << "]\n";
        cout << "Comparisons: " << comparisons << endl;
        cout << "Swaps: " << swaps << endl;
        cout << "Time: " << timeTaken << " seconds\n";
    }
};

int main() {
    srand(time(0));

    cout << "Enter numbers separated by spaces: ";
    vector<int> arr;
    int num;
    while (cin >> num) {
        arr.push_back(num);
        if (cin.peek() == '\n') break;
    }

    if (arr.empty()) {
        cout << "Empty input!\n";
        return 0;
    }

    QuickSort sorter;
    sorter.sortAndShow(arr, false);  // Deterministic
    sorter.sortAndShow(arr, true);   // Randomized

    return 0;
}

File: sorting.py
import random
import time

class QuickSort:
    def __init__(self):
        self.comparisons = 0
        self.swaps = 0

    def partition(self, arr, low, high, randomized=False):
        if randomized:
            r = random.randint(low, high)
            arr[r], arr[high] = arr[high], arr[r]
            self.swaps += 1

        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            self.comparisons += 1
            if arr[j] <= pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
                self.swaps += 1
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        self.swaps += 1
        return i + 1

    def quicksort(self, arr, low, high, randomized=False):
        if low < high:
            pi = self.partition(arr, low, high, randomized)
            self.quicksort(arr, low, pi - 1, randomized)
            self.quicksort(arr, pi + 1, high, randomized)

    def sort_and_analyze(self, arr, randomized=False):
        arr_copy = arr[:]
        self.comparisons = self.swaps = 0
        start = time.time()
        self.quicksort(arr_copy, 0, len(arr_copy) - 1, randomized)
        end = time.time()
        return arr_copy, self.comparisons, self.swaps, end - start

def main():
    arr = list(map(int, input("Enter numbers separated by spaces: ").split()))
    if not arr:
        print("Empty array!")
        return

    sorter = QuickSort()

    # Deterministic quicksort
    sorted_arr, comp, swp, t = sorter.sort_and_analyze(arr, False)
    print("\nDETERMINISTIC QUICKSORT")
    print("Sorted array:", sorted_arr)
    print(f"Comparisons: {comp}\nSwaps: {swp}\nTime: {t:.6f} s")

    # Randomized quicksort
    sorted_arr, comp, swp, t = sorter.sort_and_analyze(arr, True)
    print("\nRANDOMIZED QUICKSORT")
    print("Sorted array:", sorted_arr)
    print(f"Comparisons: {comp}\nSwaps: {swp}\nTime: {t:.6f} s")

if __name__ == "__main__":
    main()
